src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2370 @@
+    return 1; // limit is reached. Stop.
+  }
+
+  /* If circpad_max_global_padding_pct is non-zero, check our global
+   * process limit. */
+  if (circpad_global_max_padding_percent &&
+      circpad_global_padding_sent >= circpad_global_allowed_cells &&
+      (100*circpad_global_padding_sent) /
+         (circpad_global_padding_sent + circpad_global_nonpadding_sent) >
+         circpad_global_max_padding_percent) {
+    return 1; // global limit reached. Stop.
+  }
+
+  return 0; // All good!
+}
+
+/**
+ * Schedule the next padding time according to the machineinfo on a
+ * circuit.
+ *
+ * The histograms represent inter-packet-delay. Whenever you get an packet
+ * event you should be scheduling your next timer (after cancelling any old
+ * ones and updating tokens accordingly).
+ *
+ * Returns 1 if we decide to transition states (due to infinity bin),
+ * 0 otherwise.
+ */
+circpad_decision_t
+circpad_machine_schedule_padding(circpad_machineinfo_t *mi)
+{
+  circpad_delay_t in_us = 0;
+  struct timeval timeout;
+  tor_assert(mi);
+
+  // Don't pad in end (but  also don't cancel any previously
+  // scheduled padding either).
+  if (mi->current_state == CIRCPAD_STATE_END) {
+    log_fn(LOG_INFO, LD_CIRC, "Padding end state");
+    return CIRCPAD_STATE_UNCHANGED;
+  }
+
+  /* Check our padding limits */
+  if (circpad_machine_reached_padding_limit(mi)) {
+   if (CIRCUIT_IS_ORIGIN(mi->on_circ)) {
+      log_fn(LOG_INFO, LD_CIRC,
+           "Padding machine has reached padding limit on circuit %u",
+             TO_ORIGIN_CIRCUIT(mi->on_circ)->global_identifier);
+    } else {
+      log_fn(LOG_INFO, LD_CIRC,
+           "Padding machine has reached padding limit on circuit %"PRIu64
+           ", %d",
+           mi->on_circ->n_chan ? mi->on_circ->n_chan->global_identifier : 0,
+           mi->on_circ->n_circ_id);
+    }
+    return CIRCPAD_STATE_UNCHANGED;
+  }
+
+  if (mi->padding_timer_scheduled) {
+    /* Cancel current timer (if any) */
+    timer_disable(mi->padding_timer);
+    mi->padding_timer_scheduled = 0;
+  }
+
+  /* in_us = in microseconds */
+  in_us = circpad_machine_sample_delay(mi);
+  mi->padding_scheduled_at_us = monotime_absolute_usec();
+  log_fn(LOG_INFO,LD_CIRC,"\tPadding in %u usec", in_us);
+
+  // Don't schedule if we have infinite delay.
+  if (in_us == CIRCPAD_DELAY_INFINITE) {
+    return circpad_internal_event_infinity(mi);
+  }
+
+  if (mi->state_length == 0) {
+    /* If we're at length 0, that means we hit 0 after sending
+     * a cell earlier, and emitted an event for it, but
+     * for whatever reason we did not decide to change states then.
+     * So maybe the machine is waiting for bins empty, or for an
+     * infinity event later? That would be a strange machine,
+     * but there's no reason to make it impossible. */
+    return CIRCPAD_STATE_UNCHANGED;
+  }
+
+  if (in_us <= 0) {
+    return circpad_send_padding_cell_for_callback(mi);
+  }
+
+  timeout.tv_sec = in_us/TOR_USEC_PER_SEC;
```

Comment:
Is this 32-bit math or 64-bit math?  We need to be careful of doing frequent 64-bit division on 32-bit architectures; it's really slow.

Reply:
Hmm, I'm not sure. `in_usec` seems to be a uint32_t in the lines below:

```
  timeout.tv_sec = in_usec/TOR_USEC_PER_SEC;
  timeout.tv_usec = (in_usec%TOR_USEC_PER_SEC);
```
And here is the assembly that gets generated for the following two lines:

```
mov    $0x431bde83,%r8d
mov    %ebp,%eax
mov    %ebp,%r9d
mov    $0x6,%edi
mul    %r8d
lea    0x269b56(%rip),%rcx        # 0x555555b713e7
mov    $0x400,%esi
mov    %edx,%r8d
lea    0x269ea0(%rip),%rdx        # 0x555555b71740 <__FUNCTION__.26195>
shr    $0x12,%r8d
mov    %r8d,%eax
mov    %rax,(%rsp)

imul   $0xf4240,%r8d,%eax
sub    %eax,%r9d
mov    %r9d,%eax
mov    %rax,0x8(%rsp)
```

So it seems to be implementing the division using right shifts `shr` and a multiplication `mul`? But how does that answer your question?

Reply:
If both arguments are 32-bit values, it should only be using 32-bit math.  So, cool.  I think I probably just wasn't looking at the definition of circuitpad_delay_t when I wrote that question.